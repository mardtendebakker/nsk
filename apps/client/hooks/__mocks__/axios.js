const axios = jest.createMockFromModule('axios');

axios.create = () => ({
  get: () => {},
  post: () => {},
  put: () => {},
  delete: () => {},
  patch: () => {},
  interceptors: {
    request: {
      use: () => {},
    },
    response: {
      use: () => {},
    },
  },
});

axios.CancelToken = {
  source: () => ({
    cancel: () => {},
    token: 'token',
  }),
};

module.exports = axios;
