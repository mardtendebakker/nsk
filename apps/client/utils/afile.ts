import { AFile } from './axios/models/aFile';

export function buildAFileLink(afile: AFile): string {
  return `https://${afile.unique_server_filename}.s3.amazonaws.com/${afile.discr}/${afile.original_client_filename}`;
}
