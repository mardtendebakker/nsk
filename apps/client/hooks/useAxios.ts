import { useEffect, useRef, useState } from 'react';
import { useSnackbar } from 'notistack';
import { trans } from 'itranslator';
import axios, {
  AxiosResponse, AxiosPromise, CancelToken, AxiosError, CanceledError,
} from '../utils/axios';
import { hideProgress, showProgress } from '../components/topLinearProgress';

export const GET = 'get';
export const POST = 'post';
export const PATCH = 'patch';
export const DELETE = 'delete';
export const PUT = 'put';

type Method = 'get' | 'post' | 'patch' | 'delete' | 'put';
type ResponseType = 'json' | 'blob';

const useAxios = <Data>(
  method: Method,
  path: string,
  {
    withProgressBar = false,
    showErrorMessage = true,
    showSuccessMessage = false,
    customSuccessMessage,
    defaultParams = {},
    customStatusesMessages,
  }
  : {
    withProgressBar? : boolean,
    showErrorMessage? : boolean,
    showSuccessMessage? : boolean,
    customSuccessMessage?: (response: AxiosResponse) => string
    defaultParams?: object,
    customStatusesMessages?: { [key: number]: string }
  } = {},
) : {
    data?: Data,
    response?: AxiosResponse,
    error?: AxiosError,
    performing: boolean,
    call: Call,
  } => {
  const [response, setResponse] = useState<AxiosResponse>();
  const [error, setError] = useState<AxiosError>();
  const [performing, setPerforming] = useState<boolean>(false);
  const { enqueueSnackbar } = useSnackbar();
  const source = useRef(CancelToken.source());

  useEffect(() => source.current.cancel, []);

  async function call(
    explicitPath: string,
    params?: object,
    body?: object,
    responseType?: ResponseType,
    headers?: object,
  )
    : AxiosPromise<AxiosResponse> {
    if (method === POST || method === PATCH || method === PUT) {
      return axios[method](explicitPath, body, {
        cancelToken: source.current.token,
        params: { ...defaultParams, ...params },
        responseType,
        headers,
      });
    }

    return axios[method](explicitPath, {
      cancelToken: source.current.token, params: { ...defaultParams, ...params }, data: body, responseType,
    });
  }

  function handleSuccess(handledResponse: AxiosResponse) {
    enqueueSnackbar(
      customSuccessMessage ? customSuccessMessage(handledResponse) : trans(handledResponse.status.toString()),
      { variant: 'success' },
    );
  }

  function handleError(e: Error | AxiosError) {
    const errorResponse = (e instanceof AxiosError && e.response) ? e.response : { status: 500, data: { message: undefined } };
    const errorMessage = customStatusesMessages && customStatusesMessages[errorResponse.status];
    let messages: string[] = [];

    switch (true) {
      case typeof errorMessage == 'string':
        messages.push(errorMessage);
        break;
      case Array.isArray(errorResponse.data.message):
        messages = errorResponse.data.message;
        break;
      case typeof errorResponse.data.message == 'string' && !!errorResponse.data.message:
        messages.push(errorResponse.data.message);
        break;
      default:
        messages.push(trans(errorResponse.status.toString()));
    }

    messages.forEach((message) => {
      enqueueSnackbar(message, { variant: 'error' });
    });
  }

  return {
    data: response?.data,
    response,
    error,
    performing,
    /** @throws {Error} */
    call: async (
      {
        params, body, path: explicitPath, responseType, headers,
      }
      : { params?: object, body?: object, path?: string, responseType?: ResponseType, headers?: object }
      = {},
      cb?: (e: Error, axiosResponse?: AxiosResponse) => void,
    ): Promise<AxiosResponse | void> => {
      try {
        const finalPath = explicitPath || path;

        setPerforming(true);
        if (withProgressBar) {
          showProgress();
        }

        const resp = await call(finalPath, params, body, responseType, headers);
        setResponse(resp);

        if (showSuccessMessage || customSuccessMessage) {
          handleSuccess(resp);
        }

        if (cb) {
          cb(null, resp);
        }

        return resp;
      } catch (e) {
        if (e instanceof CanceledError) {
          throw new Error(explicitPath + path);
        }

        setError(e);
        if (showErrorMessage) {
          handleError(e);
        }

        if (cb) {
          cb(e);
        } else {
          throw e;
        }
      } finally {
        setPerforming(false);

        if (withProgressBar) {
          hideProgress();
        }
      }

      return undefined;
    },
  };
};

export type Call = (
  arg0?: { params?: object, body?: object, path?: string, responseType?: ResponseType, headers?: object },
  cb?: (e: Error, axiosResponse?: AxiosResponse) => void,
) => Promise<AxiosResponse | void>;

export default useAxios;
