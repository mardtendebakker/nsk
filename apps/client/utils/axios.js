import axios from 'axios';

export { AxiosPromise, AxiosResponse, AxiosError } from 'axios';

const baseURL = process.env.NODE_ENV === 'development' ? 'http://localhost:3333' : 'production_url';

export const client = axios.create({
  baseURL,
  timeout: 120000,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

export const { CancelToken } = axios;

export const SUPLIERS_PATH = '';

export default client;
