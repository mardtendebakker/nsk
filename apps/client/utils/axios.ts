import axios from 'axios';

export { AxiosError } from 'axios';
export type { AxiosPromise, AxiosResponse } from 'axios';

const baseURL = process.env.NODE_ENV === 'development' ? 'http://localhost:3333/api' : process.env.NX_AXIOS_BASE_URL;

export const client = axios.create({
  baseURL,
  timeout: 120000,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

export const { CancelToken } = axios;

export const CUSTOMERS_PATH = '/customers/:id';
export const SUPPLIERS_PATH = '/suppliers/:id';

export default client;

export interface Customer {
  id: number,
  name?: string,
  representative?: string,
  kvk_nr?: string,
  email?: string,
  phone?: string,
  phone2?: string,
  street?: string,
  street_extra?: string,
  city?: string,
  country?: string,
  state?: string,
  zip?: string,
  street2?: string,
  street_extra2?: string,
  city2?: string,
  country2?: string,
  state2?: string,
  zip2?: string,
  is_partner?: number,
}

export interface Supplier {
  id: number,
  name?: string,
  representative?: string,
  email?: string,
  phone?: string,
  phone2?: string,
  street?: string,
  street_extra?: string,
  city?: string,
  country?: string,
  state?: string,
  zip?: string,
  street2?: string,
  street_extra2?: string,
  city2?: string,
  country2?: string,
  state2?: string,
  zip2?: string,
  partner?: string,
}
