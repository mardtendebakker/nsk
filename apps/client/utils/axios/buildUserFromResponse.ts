import { AxiosResponse } from 'axios';
import { User } from '../../stores/security/types';

export default (response: AxiosResponse): User => ({
  username: response.data.idToken.payload['cognito:username'],
  email: response.data.idToken.payload.email,
  accessToken: response.data.idToken.jwtToken,
  refreshToken: response.data.refreshToken.token,
  emailVerified: response.data.idToken.payload.email_verified,
});
