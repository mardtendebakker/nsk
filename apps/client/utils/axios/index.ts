import axios from 'axios';
import axiosClient from './client';
import refreshToken from './refreshTokenInterceptor';

export {
  CUSTOMERS_PATH,
  SUPPLIERS_PATH,
  SIGN_IN_PATH,
  REFRESH_TOKEN_PATH,
  USER_INFO_PATH,
  SIGN_UP_PATH,
  CONFIRM_ACCOUNT_PATH,
  CHANGE_PASSWORD_PATH,
  SEND_VERIFICATION_CODE_PATH,
  FORGOT_PASSWORD_PATH,
  SALES_ORDERS_PATH,
  PURCHASE_ORDERS_PATH,
  STOCK_PRODUCTS_PATH,
  USERS_PATH,
  STOCK_REPAIR_SERVICES_PATH,
  ORDER_STATUSES_PATH,
  COMPANIES_PATH,
  LOCATIONS_PATH,
  BULK_PRINT_SALES_PATH,
  BULK_PRINT_PURCHASES_PATH,
  ORDERS_ANALYTICS_PATH,
} from './paths';
export { AxiosError } from 'axios';
export type { AxiosPromise, AxiosResponse } from 'axios';

axiosClient.interceptors.response.use((response) => response, refreshToken);

export const { CancelToken } = axios;

export default axiosClient;
