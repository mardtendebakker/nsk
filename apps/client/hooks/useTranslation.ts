import { useEffect, useState } from 'react';
import { trans, setConfig } from 'itranslator';
import { setDefaultOptions } from 'date-fns';
import {
  nl, enUS, de, fr, ar, faIR,
} from 'date-fns/locale';
import Config from 'itranslator/lib/type/config';
import { getDefaultLocale, setDefaultLocale } from '../utils/storage';
import EventEmitter from '../utils/eventEmitter';
import nlSource from '../public/translations/nl';
import enSource from '../public/translations/en';
import deSource from '../public/translations/de';
import frSource from '../public/translations/fr';
import arSource from '../public/translations/ar';
import faSource from '../public/translations/fa';

export const LOCALE_UPDATED = 'LOCALE_UPDATED';
export type Trans = (key: string, config?: Config | undefined) => string;

export const localeMapping = {
  nl: nlSource,
  en: enSource,
  de: deSource,
  ar: arSource,
  fr: frSource,
  fa: faSource,
};

export const dateFnsMapping = {
  nl,
  en: enUS,
  de,
  fr,
  ar,
  fa: faIR,
};

class LocaleEmitter extends EventEmitter {
  #locale = getDefaultLocale() || 'nl';

  constructor() {
    super();
    setDefaultOptions({ locale: dateFnsMapping[this.#locale] });
  }

  get locale() { return this.#locale; }

  emit(event: string, payload: string) {
    if (LOCALE_UPDATED === event) {
      this.#locale = payload;

      setConfig({ source: localeMapping[payload] });
      setDefaultOptions({ locale: dateFnsMapping[payload] });
      setDefaultLocale(payload);
    }

    return super.emit(event, payload);
  }
}

export const localeStore = new LocaleEmitter();

const useTranslation = (): { updateLocale: (locale: string)=> void, trans: Trans, locale: string } => {
  const [locale, setLocale] = useState<string>(localeStore.locale);

  useEffect(() => {
    localeStore.on(LOCALE_UPDATED, setLocale);
    return () => {
      localeStore.removeListener(LOCALE_UPDATED, setLocale);
    };
  }, [setLocale]);

  const updateLocale = (newLocale: string) => {
    localeStore.emit(LOCALE_UPDATED, newLocale);
    setLocale(newLocale);
  };

  return {
    updateLocale,
    trans,
    locale,
  };
};

export default useTranslation;
