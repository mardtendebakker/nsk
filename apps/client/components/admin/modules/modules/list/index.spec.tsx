import {
  render, screen, fireEvent, waitFor,
} from '@testing-library/react';
import ListContainer from '.';
import useAxios from '../../../../../hooks/useAxios';
import useTranslation from '../../../../../hooks/useTranslation';
import { PAYMENTS_FREE_TRIAL_PATH, MODULES_PATH, MODULES_CONFIGS_PATH } from '../../../../../utils/axios';

jest.mock('../../../../../hooks/useAxios', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('../../../../../hooks/useTranslation', () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock('./list', () => ({
  __esModule: true,
  default: function MockList({
    modules, onFreeTrial, onSettings, disabled,
  }) {
    return (
      <div data-testid="mock-list">
        {modules.map((module) => (
          <div key={module.id} data-testid={`module-${module.id}`}>
            <button onClick={() => onFreeTrial(module)} disabled={disabled} data-testid={`free-trial-${module.id}`} type="button">
              Free Trial
            </button>
            <button onClick={() => onSettings(module)} disabled={disabled} data-testid={`settings-${module.id}`} type="button">
              Settings
            </button>
          </div>
        ))}
      </div>
    );
  },
}));

jest.mock('./settingsModal', () => ({
  __esModule: true,
  default: function MockSettingsModal({ onSubmit, onClose, disabled }) {
    return (
      <div data-testid="settings-modal">
        <button onClick={() => onSubmit({ config: 'test' })} disabled={disabled} data-testid="submit-settings" type="button">
          Submit
        </button>
        <button onClick={onClose} data-testid="close-modal" type="button">Close</button>
      </div>
    );
  },
}));

describe('ListContainer Component', () => {
  const mockModules = [
    { id: 1, name: 'Module 1' },
    { id: 2, name: 'Module 2' },
  ];

  const mockAxiosCall = jest.fn();
  const mockFreeTrialCall = jest.fn();
  const mockConfigCall = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    (useTranslation as jest.Mock).mockReturnValue({
      trans: (key: string) => key,
    });

    (useAxios as jest.Mock)
      .mockImplementation((method, path) => {
        if (path === MODULES_PATH.replace(':id', '')) {
          return {
            data: mockModules,
            call: mockAxiosCall,
            performing: false,
          };
        }
        if (path === PAYMENTS_FREE_TRIAL_PATH.replace(':moduleName', '')) {
          return {
            call: mockFreeTrialCall,
            performing: false,
          };
        }
        if (path === MODULES_CONFIGS_PATH.replace(':moduleName', '')) {
          return {
            call: mockConfigCall,
            performing: false,
          };
        }
      });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<ListContainer />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('fetches and renders modules on mount', () => {
    render(<ListContainer />);

    expect(mockAxiosCall).toHaveBeenCalled();
    expect(screen.getByTestId('mock-list')).toBeInTheDocument();
    mockModules.forEach((module) => {
      expect(screen.getByTestId(`module-${module.id}`)).toBeInTheDocument();
    });
  });

  it('handles free trial activation', async () => {
    render(<ListContainer />);

    fireEvent.click(screen.getByTestId('free-trial-1'));

    expect(mockFreeTrialCall).toHaveBeenCalledWith({
      path: PAYMENTS_FREE_TRIAL_PATH.replace(':id', '1'),
    });

    await waitFor(() => {
      expect(mockAxiosCall).toHaveBeenCalledTimes(2);
    });
  });

  it('handles settings modal interactions', async () => {
    render(<ListContainer />);

    fireEvent.click(screen.getByTestId('settings-1'));
    expect(screen.getByTestId('settings-modal')).toBeInTheDocument();

    fireEvent.click(screen.getByTestId('submit-settings'));

    expect(mockConfigCall).toHaveBeenCalledWith({
      path: MODULES_CONFIGS_PATH.replace(':id', '1'),
      body: { config: 'test' },
    });

    await waitFor(() => {
      expect(screen.queryByTestId('settings-modal')).not.toBeInTheDocument();
      expect(mockAxiosCall).toHaveBeenCalledTimes(2);
    });
  });

  it('closes settings modal without submitting', () => {
    render(<ListContainer />);

    fireEvent.click(screen.getByTestId('settings-1'));
    expect(screen.getByTestId('settings-modal')).toBeInTheDocument();

    fireEvent.click(screen.getByTestId('close-modal'));
    expect(screen.queryByTestId('settings-modal')).not.toBeInTheDocument();
    expect(mockConfigCall).not.toHaveBeenCalled();
  });

  it('disables interactions while performing operations', () => {
    (useAxios as jest.Mock)
      .mockImplementation((method, path) => {
        if (path === MODULES_PATH.replace(':id', '')) {
          return {
            data: mockModules,
            call: mockAxiosCall,
            performing: true,
          };
        }
        return {
          call: jest.fn(),
          performing: false,
        };
      });

    render(<ListContainer />);

    mockModules.forEach((module) => {
      expect(screen.getByTestId(`free-trial-${module.id}`)).toBeDisabled();
      expect(screen.getByTestId(`settings-${module.id}`)).toBeDisabled();
    });
  });
});
