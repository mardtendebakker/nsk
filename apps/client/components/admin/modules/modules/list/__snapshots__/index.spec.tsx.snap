// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListContainer Component matches snapshot 1`] = `
<DocumentFragment>
  <h4
    class="MuiTypography-root MuiTypography-h4 css-4il6ew-MuiTypography-root"
  >
    modules
  </h4>
  <div
    data-testid="mock-list"
  >
    <div
      data-testid="module-1"
    >
      <button
        data-testid="free-trial-1"
        type="button"
      >
        Free Trial
      </button>
      <button
        data-testid="settings-1"
        type="button"
      >
        Settings
      </button>
    </div>
    <div
      data-testid="module-2"
    >
      <button
        data-testid="free-trial-2"
        type="button"
      >
        Free Trial
      </button>
      <button
        data-testid="settings-2"
        type="button"
      >
        Settings
      </button>
    </div>
  </div>
</DocumentFragment>
`;
